sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala

```diff
@@ -16,53 +16,168 @@
+  def getRapidsShuffleManagerClass: String = {
+    findShimProvider()
+    s"com.nvidia.spark.rapids.$shimId.RapidsShuffleManager"
+  }
+
+  def getRapidsShuffleInternal: String = {
+    findShimProvider()
+    s"org.apache.spark.sql.rapids.shims.$shimId.RapidsShuffleInternalManager"
+  }
+
+  private def updateSparkClassLoader(): Unit = {
+    val contextClassLoader = Thread.currentThread().getContextClassLoader
+    Option(contextClassLoader).collect {
+      case mutable: MutableURLClassLoader => mutable
+      case replCL if replCL.getClass.getName == "org.apache.spark.repl.ExecutorClassLoader" =>
+        val parentLoaderField = replCL.getClass.getDeclaredMethod("parentLoader")
+        val parentLoader = parentLoaderField.invoke(replCL).asInstanceOf[ParentClassLoader]
+        parentLoader.getParent.asInstanceOf[MutableURLClassLoader]
+    }.foreach { mutable =>
+      // MutableURLClassloader dedupes for us
+      pluginClassLoader = contextClassLoader
+      mutable.addURL(shimURL)
+      mutable.addURL(shimCommonURL)
+    }
+  }
 
-  private def detectShimProvider(): SparkShimServiceProvider = {
+  private def getShimClassLoader(): ClassLoader = {
+    if (shimURL == null) {
+      findShimProvider()
+    }
+    if (pluginClassLoader == null) {
+      updateSparkClassLoader()
+    }
+    if (pluginClassLoader == null) {
+      if (tmpClassLoader == null) {
+        tmpClassLoader = new MutableURLClassLoader(Array(shimURL, shimCommonURL),
+          getClass.getClassLoader)
+      }
+      tmpClassLoader
+    } else {
+      pluginClassLoader
+    }
+  }
+
+  private def detectShimProvider(): String = {
     val sparkVersion = getSparkVersion
     logInfo(s"Loading shim for Spark version: $sparkVersion")
 
-    // This is not ideal, but pass the version in here because otherwise loader that match the
-    // same version (3.0.1 Apache and 3.0.1 Databricks) would need to know how to differentiate.
-    val sparkShimLoaders = ServiceLoader.load(classOf[SparkShimServiceProvider])
-        .asScala.filter(_.matchesVersion(sparkVersion))
-    if (sparkShimLoaders.size > 1) {
-      throw new IllegalArgumentException(s"Multiple Spark Shim Loaders found: $sparkShimLoaders")
+    // Emulating service loader manually because we have a non-standard jar layout for classes
+    val serviceProviderListPath = "/META-INF/services/" + classOf[SparkShimServiceProvider].getName
+    val serviceProviderList = Option(getClass.getResource(serviceProviderListPath)).toSeq
+      .map(scala.io.Source.fromURL)
+      .flatMap(_.getLines())
+
+    val shimServiceProviderOpt = serviceProviderList.flatMap { shimServiceProviderStr =>
+      val mask = shimIdFromPackageName(shimServiceProviderStr)
+      try {
+        val shimURL = new java.net.URL(s"${shimRootURL.toString}$mask/")
+        val shimClassLoader = new MutableURLClassLoader(Array(shimURL, shimCommonURL),
+          getClass.getClassLoader)
+        // can't use ServiceLoader with parallel world layout
```

Comment:
what does this mean?

Reply:
Even if I pass the modified classloader to the  ServiceLoader.load invocation it expects META-INF/services (which is ok) and provider classes in the normal root / entry (not ok)